compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,105 @@ static bool consumesInputOp(NamedSequenceOp op) {
   return false;
 }
 
+static bool hasConsumedArgument(ForeachMatchOp op) {
+  Value operand = op->getOperand(0);
+  if (auto blockArg = dyn_cast<BlockArgument>(operand)) {
+    Operation *parentOp = blockArg.getOwner()->getParentOp();
+    if (auto namedSequenceOp = dyn_cast<NamedSequenceOp>(parentOp)) {
+      return namedSequenceOp.getArgAttr(blockArg.getArgNumber(),
+                                        kArgConsumedAttrName) != nullptr;
+    }
+  }
+  return false;
+}
+
+// Extracts the base name from a `specName`.
+// - If `specName` ends with `_<number>`, the base name is everything before
+// `_`.
+// - Otherwise, the full name is used.
+static std::string getBaseName(StringRef specName) {
+  auto pos = specName.rfind('_');
+  if (pos != StringRef::npos) {
+    StringRef potentialBase = specName.substr(0, pos);
+    unsigned suffix;
+    if (!specName.substr(pos + 1).getAsInteger(10, suffix)) {
+      return potentialBase.str(); // If valid number, return base name
+    }
+  }
+  return specName.str(); // Otherwise, return full name
+}
+
+static std::string
+getUniqueSpecName(StringRef specName,
+                  llvm::StringMap<unsigned> &specNameCounts) {
+  std::string specBaseName = getBaseName(specName);
+  unsigned specNameSeenCount = specNameCounts[specBaseName]++;
+
+  return (specNameSeenCount == 0)
+             ? specBaseName
+             : llvm::formatv("{}_{}", specBaseName, specNameSeenCount).str();
+}
+
+// Collects `NamedSequenceOp`s and `ForeachMatchOp`s from a given module.
+// - Moves `NamedSequenceOp`s to `namedSequenceOpsToMove` if their name is not
+//   `__kernel_config` or they do not have
+//   the`iree_codegen.tuning_spec_entrypoint` attribute.
+// - Maps `NamedSequenceOp`s that are used as matchers or actions to their
+//   corresponding `ForeachMatchOp` in `namedSequenceToForeachMatch`.
+static void collectNamedSequenceAndForeachMatchOpsFromModule(
+    ModuleOp module, SmallVector<NamedSequenceOp> &namedSequenceOpsToMove,
+    SmallVector<ForeachMatchOp> &foreachMatchOps,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp>
+        &namedSequenceToForeachMatch) {
```

Comment:
Could you define a struct with this info that can be directly returned. This should improve readability.